File: app/src/main/java/com/einkreader/core/storage/DatabaseHelper.java
package com.einkreader.core.storage;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.os.Build;

import com.einkreader.ui.reader.DebugLog;
import android.util.LruCache;

import java.util.ArrayList;
import java.util.List;

/**
 * DatabaseHelper —— 轻量级 SQLite 存储实现
 *
 * 设计要点：
 * 1. 继承 SQLiteOpenHelper，兼容 Android 4.4（API 19），无需 Room
 * 2. 两张表：books（书籍元信息）、progress（阅读进度与统计）
 * 3. 使用 fileKey（文件名+大小+修改时间的指纹）作为主键，避免重复入库
 * 4. 所有方法均为线程安全的同步实现（单机阅读无需高并发）
 */
public class DatabaseHelper extends SQLiteOpenHelper implements BookStorage {

    private static final String TAG = "DatabaseHelper";

    private static final String DB_NAME = "einkreader.db";
    private static final int DB_VERSION = 1;

    private static final String TABLE_BOOKS = "books";
    private static final String TABLE_PROGRESS = "progress";

    private static final String KEY_FILE_KEY = "file_key";
    private static final String KEY_FILE_PATH = "file_path";
    private static final String KEY_TITLE = "title";
    private static final String KEY_FORMAT = "format";
    private static final String KEY_FILE_SIZE = "file_size";
    private static final String KEY_LAST_MODIFIED = "last_modified";
    private static final String KEY_ADDED_AT = "added_at";
    private static final String KEY_TOTAL_CHAPTERS = "total_chapters";
    private static final String KEY_LAST_READ_TIME = "last_read_time";
    private static final String KEY_TOTAL_READ_MS = "total_read_ms";

    private static final String KEY_CHAPTER_INDEX = "chapter_index";
    private static final String KEY_PAGE_INDEX = "page_index";
    private static final String KEY_UPDATED_AT = "updated_at";

    private Context context;
    private volatile boolean initialized = false;
    private final LruCache<String, BookProgress> progressCache = new LruCache<>(64);

    public DatabaseHelper(Context context) {
        super(context.getApplicationContext(), DB_NAME, null, DB_VERSION);
        this.context = context.getApplicationContext();
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        String createBooks = "CREATE TABLE IF NOT EXISTS " + TABLE_BOOKS + " ("
                + KEY_FILE_KEY + " TEXT PRIMARY KEY,"
                + KEY_FILE_PATH + " TEXT NOT NULL,"
                + KEY_TITLE + " TEXT,"
                + KEY_FORMAT + " TEXT,"
                + KEY_FILE_SIZE + " INTEGER DEFAULT 0,"
                + KEY_LAST_MODIFIED + " INTEGER DEFAULT 0,"
                + KEY_ADDED_AT + " INTEGER DEFAULT 0,"
                + KEY_TOTAL_CHAPTERS + " INTEGER DEFAULT 0,"
                + KEY_LAST_READ_TIME + " INTEGER DEFAULT 0,"
                + KEY_TOTAL_READ_MS + " INTEGER DEFAULT 0"
                + ")";

        String createProgress = "CREATE TABLE IF NOT EXISTS " + TABLE_PROGRESS + " ("
                + KEY_FILE_KEY + " TEXT PRIMARY KEY,"
                + KEY_CHAPTER_INDEX + " INTEGER DEFAULT 0,"
                + KEY_PAGE_INDEX + " INTEGER DEFAULT 0,"
                + KEY_TOTAL_CHAPTERS + " INTEGER DEFAULT 0,"
                + KEY_UPDATED_AT + " INTEGER DEFAULT 0"
                + ")";

        db.execSQL(createBooks);
        db.execSQL(createProgress);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        if (oldVersion < 1) {
            db.execSQL("CREATE TABLE IF NOT EXISTS " + TABLE_BOOKS + " ("
                    + KEY_FILE_KEY + " TEXT PRIMARY KEY,"
                    + KEY_FILE_PATH + " TEXT NOT NULL,"
                    + KEY_TITLE + " TEXT,"
                    + KEY_FORMAT + " TEXT,"
                    + KEY_FILE_SIZE + " INTEGER DEFAULT 0,"
                    + KEY_LAST_MODIFIED + " INTEGER DEFAULT 0,"
                    + KEY_ADDED_AT + " INTEGER DEFAULT 0,"
                    + KEY_TOTAL_CHAPTERS + " INTEGER DEFAULT 0,"
                    + KEY_LAST_READ_TIME + " INTEGER DEFAULT 0,"
                    + KEY_TOTAL_READ_MS + " INTEGER DEFAULT 0"
                    + ")");
            db.execSQL("CREATE TABLE IF NOT EXISTS " + TABLE_PROGRESS + " ("
                    + KEY_FILE_KEY + " TEXT PRIMARY KEY,"
                    + KEY_CHAPTER_INDEX + " INTEGER DEFAULT 0,"
                    + KEY_PAGE_INDEX + " INTEGER DEFAULT 0,"
                    + KEY_TOTAL_CHAPTERS + " INTEGER DEFAULT 0,"
                    + KEY_UPDATED_AT + " INTEGER DEFAULT 0"
                    + ")");
        }
        // 未来升级在此处按 oldVersion 分段处理
    }

    @Override
    public void onConfigure(SQLiteDatabase db) {
        super.onConfigure(db);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
            db.enableWriteAheadLogging();
        } else {
            db.execSQL("PRAGMA journal_mode=WAL");
        }
    }

    @Override
    public void initialize() {
        if (initialized) return;
        try {
            getWritableDatabase(); // 触发 onCreate 或 onUpgrade
            initialized = true;
            DebugLog.log(TAG, "Database initialized");
        } catch (Exception e) {
            DebugLog.log(TAG, "Database init failed: " + e.getMessage());
        }
    }

    // ==================== BookStorage 实现 ====================

    @Override
    public synchronized void upsertBook(BookRecord record) {
        if (record == null || record.fileKey == null) return;
        try {
            SQLiteDatabase db = getWritableDatabase();
            ContentValues cv = new ContentValues();
            cv.put(KEY_FILE_KEY, record.fileKey);
            cv.put(KEY_FILE_PATH, record.filePath);
            cv.put(KEY_TITLE, record.title);
            cv.put(KEY_FORMAT, record.format);
            cv.put(KEY_FILE_SIZE, record.fileSize);
            cv.put(KEY_LAST_MODIFIED, record.lastModified);
            cv.put(KEY_ADDED_AT, record.addedAt > 0 ? record.addedAt : System.currentTimeMillis());
            cv.put(KEY_TOTAL_CHAPTERS, record.totalChapters);
            cv.put(KEY_LAST_READ_TIME, record.lastReadTime);
            cv.put(KEY_TOTAL_READ_MS, record.totalReadMs);
            db.insertWithOnConflict(TABLE_BOOKS, null, cv, SQLiteDatabase.CONFLICT_REPLACE);
        } catch (Exception e) {
            DebugLog.log(TAG, "upsertBook failed: " + e.getMessage());
        }
    }

    @Override
    public synchronized BookRecord getBook(String fileKey) {
        if (fileKey == null) return null;
        Cursor c = null;
        try {
            SQLiteDatabase db = getReadableDatabase();
            c = db.query(TABLE_BOOKS, null, KEY_FILE_KEY + "=?", new String[]{fileKey}, null, null, null);
            if (c != null && c.moveToFirst()) {
                return cursorToBookRecord(c);
            }
        } catch (Exception e) {
            DebugLog.log(TAG, "getBook failed: " + e.getMessage());
        } finally {
            if (c != null) c.close();
        }
        return null;
    }

    @Override
    public synchronized List<BookRecord> listAllBooks() {
        List<BookRecord> result = new ArrayList<BookRecord>();
        Cursor c = null;
        try {
            SQLiteDatabase db = getReadableDatabase();
            c = db.query(TABLE_BOOKS, null, null, null, null, null,
                    KEY_LAST_READ_TIME + " DESC, " + KEY_ADDED_AT + " DESC");
            if (c != null && c.moveToFirst()) {
                do {
                    result.add(cursorToBookRecord(c));
                } while (c.moveToNext());
            }
        } catch (Exception e) {
            DebugLog.log(TAG, "listAllBooks failed: " + e.getMessage());
        } finally {
            if (c != null) c.close();
        }
        return result;
    }

    @Override
    public synchronized void deleteBook(String fileKey) {
        if (fileKey == null) return;
        progressCache.remove(fileKey);
        SQLiteDatabase db = getWritableDatabase();
        try {
            db.beginTransaction();
            db.delete(TABLE_BOOKS, KEY_FILE_KEY + "=?", new String[]{fileKey});
            db.delete(TABLE_PROGRESS, KEY_FILE_KEY + "=?", new String[]{fileKey});
            db.setTransactionSuccessful();
        } catch (Exception e) {
            DebugLog.log(TAG, "deleteBook failed: " + e.getMessage());
        } finally {
            db.endTransaction();
        }
    }

    @Override
    public synchronized void saveProgress(BookProgress progress) {
        if (progress == null || progress.fileKey == null) return;
        progressCache.put(progress.fileKey, progress);
        SQLiteDatabase db = getWritableDatabase();
        try {
            db.beginTransaction();
            ContentValues cv = new ContentValues();
            cv.put(KEY_FILE_KEY, progress.fileKey);
            cv.put(KEY_CHAPTER_INDEX, progress.chapterIndex);
            cv.put(KEY_PAGE_INDEX, progress.pageIndex);
            cv.put(KEY_TOTAL_CHAPTERS, progress.totalChapters);
            cv.put(KEY_UPDATED_AT, System.currentTimeMillis());
            db.insertWithOnConflict(TABLE_PROGRESS, null, cv, SQLiteDatabase.CONFLICT_REPLACE);

            // 同步更新 books 表的最近阅读时间
            ContentValues bcv = new ContentValues();
            bcv.put(KEY_LAST_READ_TIME, System.currentTimeMillis());
            db.update(TABLE_BOOKS, bcv, KEY_FILE_KEY + "=?", new String[]{progress.fileKey});
            db.setTransactionSuccessful();
        } catch (Exception e) {
            DebugLog.log(TAG, "saveProgress failed: " + e.getMessage());
        } finally {
            db.endTransaction();
        }
    }

    @Override
    public synchronized BookProgress loadProgress(String fileKey) {
        if (fileKey == null) return null;
        BookProgress cached = progressCache.get(fileKey);
        if (cached != null) return cached;
        Cursor c = null;
        try {
            SQLiteDatabase db = getReadableDatabase();
            c = db.query(TABLE_PROGRESS, null, KEY_FILE_KEY + "=?", new String[]{fileKey}, null, null, null);
            if (c != null && c.moveToFirst()) {
                BookProgress p = new BookProgress();
                p.fileKey = c.getString(c.getColumnIndex(KEY_FILE_KEY));
                p.chapterIndex = c.getInt(c.getColumnIndex(KEY_CHAPTER_INDEX));
                p.pageIndex = c.getInt(c.getColumnIndex(KEY_PAGE_INDEX));
                p.totalChapters = c.getInt(c.getColumnIndex(KEY_TOTAL_CHAPTERS));
                p.updatedAt = c.getLong(c.getColumnIndex(KEY_UPDATED_AT));
                progressCache.put(fileKey, p);
                return p;
            }
        } catch (Exception e) {
            DebugLog.log(TAG, "loadProgress failed: " + e.getMessage());
        } finally {
            if (c != null) c.close();
        }
        return null;
    }

    @Override
    public synchronized void addReadTime(String fileKey, long deltaMs) {
        if (fileKey == null || deltaMs <= 0) return;
        try {
            SQLiteDatabase db = getWritableDatabase();
            // SQLite 不支持原生累加，用 SQL 表达式
            db.execSQL("UPDATE " + TABLE_BOOKS + " SET " + KEY_TOTAL_READ_MS
                    + " = " + KEY_TOTAL_READ_MS + " + ?," + KEY_LAST_READ_TIME + " = ? WHERE "
                    + KEY_FILE_KEY + " = ?",
                    new Object[]{deltaMs, System.currentTimeMillis(), fileKey});
        } catch (Exception e) {
            DebugLog.log(TAG, "addReadTime failed: " + e.getMessage());
        }
    }

    @Override
    public synchronized long getTotalReadTime() {
        Cursor c = null;
        try {
            SQLiteDatabase db = getReadableDatabase();
            c = db.rawQuery("SELECT SUM(" + KEY_TOTAL_READ_MS + ") AS total FROM " + TABLE_BOOKS, null);
            if (c != null && c.moveToFirst()) {
                return c.getLong(0);
            }
        } catch (Exception e) {
            DebugLog.log(TAG, "getTotalReadTime failed: " + e.getMessage());
        } finally {
            if (c != null) c.close();
        }
        return 0L;
    }

    // ==================== 内部工具 ====================

    private BookRecord cursorToBookRecord(Cursor c) {
        BookRecord r = new BookRecord();
        r.fileKey = c.getString(c.getColumnIndex(KEY_FILE_KEY));
        r.filePath = c.getString(c.getColumnIndex(KEY_FILE_PATH));
        r.title = c.getString(c.getColumnIndex(KEY_TITLE));
        r.format = c.getString(c.getColumnIndex(KEY_FORMAT));
        r.fileSize = c.getLong(c.getColumnIndex(KEY_FILE_SIZE));
        r.lastModified = c.getLong(c.getColumnIndex(KEY_LAST_MODIFIED));
        r.addedAt = c.getLong(c.getColumnIndex(KEY_ADDED_AT));
        r.totalChapters = c.getInt(c.getColumnIndex(KEY_TOTAL_CHAPTERS));
        r.lastReadTime = c.getLong(c.getColumnIndex(KEY_LAST_READ_TIME));
        r.totalReadMs = c.getLong(c.getColumnIndex(KEY_TOTAL_READ_MS));
        return r;
    }
}

File: app/src/main/java/com/einkreader/core/storage/DatabaseHelper.kt
package com.einkreader.core.storage

import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteDatabase.CONFLICT_REPLACE
import android.util.LruCache
import com.einkreader.ui.reader.DebugLog
import java.util.ArrayList

class DatabaseHelper(
    context: Context
) : android.database.sqlite.SQLiteOpenHelper(
        context.applicationContext, "einkreader.db", null, 1
    ), BookStorage {

    companion object {
        private const val TAG = "DatabaseHelper"
        private const val KEY_FILE_KEY = "file_key"
        private const val KEY_FILE_PATH = "file_path"
        private const val KEY_TITLE = "title"
        private const val KEY_FORMAT = "format"
        private const val KEY_FILE_SIZE = "file_size"
        private const val KEY_LAST_MODIFIED = "last_modified"
        private const val KEY_ADDED_AT = "added_at"
        private const val KEY_TOTAL_CHAPTERS = "total_chapters"
        private const val KEY_LAST_READ_TIME = "last_read_time"
        private const val KEY_TOTAL_READ_MS = "total_read_ms"
        private const val KEY_CHAPTER_INDEX = "chapter_index"
        private const val KEY_PAGE_INDEX = "page_index"
        private const val KEY_UPDATED_AT = "updated_at"
    }

    private var initialized = false
    private val progressCache = LruCache<String, BookStorage.BookProgress>(64)

    override fun onConfigure(db: SQLiteDatabase) {
        super.onConfigure(db)
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.JELLY_BEAN) {
            db.enableWriteAheadLogging()
        } else {
            db.execSQL("PRAGMA journal_mode=WAL")
        }
    }

    override fun onCreate(db: SQLiteDatabase) {
        val createBooks = """
            CREATE TABLE IF NOT EXISTS books (
                file_key TEXT PRIMARY KEY, file_path TEXT NOT NULL,
                title TEXT, format TEXT, file_size INTEGER DEFAULT 0,
                last_modified INTEGER DEFAULT 0, added_at INTEGER DEFAULT 0,
                total_chapters INTEGER DEFAULT 0, last_read_time INTEGER DEFAULT 0,
                total_read_ms INTEGER DEFAULT 0
            )
        """.trimIndent()
        val createProgress = """
            CREATE TABLE IF NOT EXISTS progress (
                file_key TEXT PRIMARY KEY, chapter_index INTEGER DEFAULT 0,
                page_index INTEGER DEFAULT 0, total_chapters INTEGER DEFAULT 0,
                updated_at INTEGER DEFAULT 0
            )
        """.trimIndent()
        db.execSQL(createBooks)
        db.execSQL(createProgress)
    }

    override fun onUpgrade(db: SQLiteDatabase?, oldVersion: Int, newVersion: Int) {
        // V1 only
    }

    override fun initialize() {
        if (initialized) return
        try {
            writableDatabase
            initialized = true
            DebugLog.log(TAG, "Database initialized")
        } catch (e: Exception) {
            DebugLog.error(TAG, "Database init failed", e)
        }
    }

    override fun upsertBook(record: BookStorage.BookRecord) {
        if (record.fileKey == null) return
        try {
            val db = writableDatabase
            val cv = android.content.ContentValues().apply {
                put(KEY_FILE_KEY, record.fileKey)
                put(KEY_FILE_PATH, record.filePath)
                put(KEY_TITLE, record.title)
                put(KEY_FORMAT, record.format)
                put(KEY_FILE_SIZE, record.fileSize)
                put(KEY_LAST_MODIFIED, record.lastModified)
                put(KEY_ADDED_AT, if (record.addedAt > 0) record.addedAt else System.currentTimeMillis())
                put(KEY_TOTAL_CHAPTERS, record.totalChapters)
                put(KEY_LAST_READ_TIME, record.lastReadTime)
                put(KEY_TOTAL_READ_MS, record.totalReadMs)
            }
            db.insertWithOnConflict("books", null, cv, CONFLICT_REPLACE)
        } catch (e: Exception) {
            DebugLog.error(TAG, "upsertBook failed", e)
        }
    }

    override fun getBook(fileKey: String): BookStorage.BookRecord? {
        if (fileKey == null) return null
        return try {
            val c = readableDatabase.query("books", null, "$KEY_FILE_KEY=?", arrayOf(fileKey), null, null, null)
            c?.use { cursor ->
                if (cursor.moveToFirst()) cursorToBookRecord(cursor) else null
            } ?: null
        } catch (e: Exception) {
            DebugLog.error(TAG, "getBook failed", e)
            null
        }
    }

    override fun listAllBooks(): List<BookStorage.BookRecord> {
        val result = ArrayList<BookStorage.BookRecord>()
        try {
            val c = readableDatabase.query(
                "books", null, null, null, null, null,
                "$KEY_LAST_READ_TIME DESC, $KEY_ADDED_AT DESC"
            )
            c?.use { cursor ->
                while (cursor.moveToNext()) result.add(cursorToBookRecord(cursor))
            }
        } catch (e: Exception) {
            DebugLog.error(TAG, "listAllBooks failed", e)
        }
        return result
    }

    override fun deleteBook(fileKey: String) {
        if (fileKey == null) return
        progressCache.remove(fileKey)
        val db = writableDatabase
        try {
            db.beginTransaction()
            db.delete("books", "$KEY_FILE_KEY=?", arrayOf(fileKey))
            db.delete("progress", "$KEY_FILE_KEY=?", arrayOf(fileKey))
            db.setTransactionSuccessful()
        } catch (e: Exception) {
            DebugLog.log(TAG, "deleteBook failed: ${e.message}")
        } finally {
            db.endTransaction()
        }
    }

    override fun saveProgress(progress: BookStorage.BookProgress) {
        if (progress.fileKey == null) return
        progressCache.put(checkNotNull(progress.fileKey), progress)
        val db = writableDatabase
        try {
            db.beginTransaction()
            val cv = android.content.ContentValues().apply {
                put(KEY_FILE_KEY, progress.fileKey)
                put(KEY_CHAPTER_INDEX, progress.chapterIndex)
                put(KEY_PAGE_INDEX, progress.pageIndex)
                put(KEY_TOTAL_CHAPTERS, progress.totalChapters)
                put(KEY_UPDATED_AT, System.currentTimeMillis())
            }
            db.insertWithOnConflict("progress", null, cv, CONFLICT_REPLACE)
            val bcv = android.content.ContentValues()
            bcv.put(KEY_LAST_READ_TIME, System.currentTimeMillis())
            db.update("books", bcv, "$KEY_FILE_KEY=?", arrayOf(checkNotNull(progress.fileKey)))
            db.setTransactionSuccessful()
        } catch (e: Exception) {
            DebugLog.log(TAG, "saveProgress failed: ${e.message}")
        } finally {
            db.endTransaction()
        }
    }

    override fun loadProgress(fileKey: String): BookStorage.BookProgress? {
        if (fileKey == null) return null
        progressCache.get(fileKey)?.let { return it }
        return try {
            val c = readableDatabase.query("progress", null, "$KEY_FILE_KEY=?", arrayOf(fileKey), null, null, null)
            c?.use { cursor ->
                if (cursor.moveToFirst()) {
                    val p = BookStorage.BookProgress().apply {
                        this.fileKey = cursor.getString(cursor.getColumnIndexOrThrow(KEY_FILE_KEY))
                        chapterIndex = cursor.getInt(cursor.getColumnIndexOrThrow(KEY_CHAPTER_INDEX))
                        pageIndex = cursor.getInt(cursor.getColumnIndexOrThrow(KEY_PAGE_INDEX))
                        totalChapters = cursor.getInt(cursor.getColumnIndexOrThrow(KEY_TOTAL_CHAPTERS))
                        updatedAt = cursor.getLong(cursor.getColumnIndexOrThrow(KEY_UPDATED_AT))
                    }
                    progressCache.put(fileKey, p)
                    p
                } else null
            } ?: null
        } catch (e: Exception) {
            DebugLog.log(TAG, "loadProgress failed: ${e.message}")
            null
        }
    }

    override fun addReadTime(fileKey: String, deltaMs: Long) {
        if (fileKey == null || deltaMs <= 0L) return
        try {
            val sql = "UPDATE books SET $KEY_TOTAL_READ_MS = $KEY_TOTAL_READ_MS + ?, $KEY_LAST_READ_TIME = ? WHERE $KEY_FILE_KEY = ?"
            writableDatabase.execSQL(sql, arrayOf<Any>(deltaMs, System.currentTimeMillis(), fileKey))
        } catch (e: Exception) {
            DebugLog.log(TAG, "addReadTime failed: ${e.message}")
        }
    }

    override fun getTotalReadTime(): Long {
        val c = readableDatabase.rawQuery("SELECT SUM($KEY_TOTAL_READ_MS) AS total FROM books", null)
        return try {
            if (c != null && c.moveToFirst()) c.getLong(0) else 0L
        } finally {
            c?.close()
        }
    }

    private fun cursorToBookRecord(c: android.database.Cursor): BookStorage.BookRecord {
        return BookStorage.BookRecord().apply {
            fileKey = c.getString(c.getColumnIndexOrThrow(KEY_FILE_KEY))
            filePath = c.getString(c.getColumnIndexOrThrow(KEY_FILE_PATH))
            title = c.getString(c.getColumnIndexOrThrow(KEY_TITLE))
            format = c.getString(c.getColumnIndexOrThrow(KEY_FORMAT))
            fileSize = c.getLong(c.getColumnIndexOrThrow(KEY_FILE_SIZE))
            lastModified = c.getLong(c.getColumnIndexOrThrow(KEY_LAST_MODIFIED))
            addedAt = c.getLong(c.getColumnIndexOrThrow(KEY_ADDED_AT))
            totalChapters = c.getInt(c.getColumnIndexOrThrow(KEY_TOTAL_CHAPTERS))
            lastReadTime = c.getLong(c.getColumnIndexOrThrow(KEY_LAST_READ_TIME))
            totalReadMs = c.getLong(c.getColumnIndexOrThrow(KEY_TOTAL_READ_MS))
        }
    }
}
